openhtf/util/__init__.py

```diff
@@ -106,6 +108,23 @@ def __get__(self, instance, owner):
     return self._func(owner)
 
 
+def safe_format(target, **kwargs):
+  """Formats a string without requiring all values to be present.
+
+  This function allows substitutions to be gradually made in several steps
+  rather than all at once.  Similar to string.Template.safe_substitute.
+  """
+  output = target[:]
+  tags = dict(re.findall(r'(\{(.*?)\})', target))
+
+  for tag, var in re.findall(r'(\{(.*?)\})', output):
+    root = var.split('.')[0]  # dot notation
+    root = root.split('[')[0]  # dict notation
+    if root in kwargs:
+      output = output.replace(tag, tag.format(**{root: kwargs[root]}))
+
+  return output
+
 def format_string(target, **kwargs):
   """Formats a string in any of three ways (or not at all).
```

Comment:
Note: util.format_string and util.safe_format don't support string.format via unnamed *args.

I think this is appropriate since it's only used by the framework on kwargs passed.

i.e. Using standard python string format, this works:
```
'{} {}'.format('foo', 'bar') # returns 'foo bar'
```

But this won't:
```
util.format_string('{} {}', 'foo', 'bar')  # TypeError: format_string() takes exactly 1 argument (3 given)
```